### ISSUE TITLE
Unable to build indexes

### ISSUE THREAD

**Issue by steveha-ziprecruiter** [foursquare/fsqio#5]
I have had no trouble building indexes from the old `twofishes` repository.  However, I have been completely unable to build indexes from the `fsqio` repo, despite multiple attempts.

* `parse.py` attempts to check MongoDB version using the path: `dependencies/mongodb/bin/mongo`
  However, I can't find any documentation that shows how to set up the `dependencies` subdirectory, or any script that sets it up.  I solved this problem by simply modifying `parse.py`, changing the filename to `mongo` and letting `parse.py` find it through the `PATH` environment variable.

* Index building failed with a JVM garbage collection error.  I attempted to solve this problem by modifying `parse.py` to add these JVM flags:

```
jvm_args.append("-Xms8G")
jvm_args.append("-Xmx8G")
jvm_args.append("-XX:+UseConcMarkSweepGC")
```

* Index building failed with MongoDB reporting a duplicate key error.

```
6611072 INFO  i.f.t.i.i.geonames.GeonamesParser - imported 10940000 features so far
6615337 INFO  i.f.t.i.i.geonames.GeonamesParser - imported 10950000 features so far
6617326 INFO  i.f.t.i.i.geonames.GeonamesParser$ - finished: parse global features in 6591 secs / 109 mins
6617326 INFO  i.f.t.i.i.geonames.GeonamesParser$ - starting: parse global postal codes
6617329 INFO  i.f.t.i.i.geonames.GeonamesParser - imported 0 postal codes so far
EException in thread "main" com.mongodb.MongoException$DuplicateKey: { "serverUsed" : "127.0.0.1:27017" , "err" : "E11000 duplicate key error index: geocoder.features.$_id_  dup key: { : 144396663052787542 }" , "code" : 11000 , "n" : 0 , "connectionId" : 2 , "ok" : 1.0}
```

I tried building with different options, I tried re-cloning `fsqio` and I made sure that the MongoDB data directory was completely empty before restarting MongoDB and the build.  I still had the duplicate key error.

* I have been trying to build on an Amazon EC2 instance, and therefore under Linux.  I am right now trying to get a Mac so I can set up my build environment to be exactly like what Foursquare uses.  (I do think it would be desirable to have Linux as a supported build platform.)

* I would very much appreciate any additional documentation you could give about index building and how to set it up.  I don't know what purpose the S2 indexes serve, I don't know for certain how to build in stages...  For example: if I build only the reverse geo indexes, and then I want to build the autocomplete indexes... can I just run the index build again with `--noreload` and the result will be indexes with both reverse geo and autocomplete?

* I understand that Foursquare uses a Hadoop cluster to build indexes.  I would very much appreciate a README file describing how to do that.  I haven't used a Hadoop cluster and I don't even know where to start.  Do I need to learn Java, Scala, and the whole Hadoop stack just to run a Hadoop build?

* I would appreciate the dependencies for building to be better documented.  For example, is Oracle JDK the only supported JDK?  (In the `twofishes` repo, OpenJDK 7 works just fine.  In the `fsqio` repo, I tried OpenJDK 8 and had mysterious problems so I installed Oracle JDK.)  Also, is a specific version of Scala required, or just "anything 2.10 or newer"?  Postgres is listed as a requirement.... is that required for index building or just for the `mongo2postgis.py` script?

I apologize if this is too much stuff for one issue in the issue tracker.  If you would prefer I ask these questions in another venue please let me know where; and if you would prefer I split this up into multiple issue tickets, just let me know.  Thank you.

**Comment by rahulpratapm** [comment]
Thanks for the report. We'll take a look.

**Comment by steveha-ziprecruiter** [comment]
Hello.  This ticket is now 12 weeks old.  I'm still hoping for an answer.

We are seeing the duplicate key error whether we build using a Mac, or Linux on EC2.

The duplicate key error is a real problem for us.  We are unable to build indexes using the "fsqio" repo, and we have been relying on the obsolete "twofishes" repo since March.

Can you give us a workaround?  If we build using Hadoop would it avoid this problem?  If so, can you please point me to any documentation about building using Hadoop?

**Comment by rahulpratapm** [comment]
Hi Steve,

I just pushed a fix for this last week. Let me know if you still see it.

Also, sorry for the delay. We haven't yet built a solid workflow around routing and tracking twofishes issues in the common repo.

Rahul.

**Comment by rahulpratapm** [comment]
Sorry, for some reason, that change seems to have not yet propagated to this repo. Am working with the internal build team on this. Will let you know when it makes it across.

**Comment by mateor** [comment]
I updated Fsq.io to include Rahul's fix for this issue. Sorry for the delay - we were blocked from updating by an internal issue. Thanks for the detailed issue, please feel free to reopen this if you continue to have problems.

**Comment by mateor** [comment]
I should have included the referenced SHA, here it is: 03ddbca2d32f9935d0cf0289d3d460315db6ef1e

**Comment by steveha-ziprecruiter** [comment]
The fix worked for me; I am able to build Twofishes.  Thank you.
